### ISSUE TITLE
Issues with Geth gas bidding: "Known transaction" followed by "Nonce too low" error loop

### ISSUE THREAD

**Issue by treevez** [mining-visualizer/MVis-ethminer#4]
Setup geth node successfully, enabled price bidding starting at 15 gas up to 50.  Below are the logs:

```
d12 11:17:16.011> Solution found; Submitting to node ...
d12 11:17:16.097> Tx hash :, gasPrice : 15
d12 11:17:17.934> Adjusting gas price to 17, new tx hash=
d12 11:17:19.331> Adjusting gas price to 19, new tx hash=
d12 11:17:19.397> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: K
nown transaction ()
d12 11:17:19.478> Trying again in 5 seconds ...
d12 11:17:24.576> Connecting to node at 127.0.0.1:8545 ...
d12 11:17:27.055> Connection established.
d12 11:17:27.111> New challenge :
d12 11:17:27.344> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: K
nown transaction ()
d12 11:17:27.406> Trying again in 5 seconds ...
d12 11:17:32.495> Connecting to node at 127.0.0.1:8545 ...
d12 11:17:32.862> Connection established.
d12 11:17:32.937> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: K
nown transaction ()
d12 11:17:33.012> Trying again in 5 seconds ...
d12 11:17:38.092> Connecting to node at 127.0.0.1:8545 ...
d12 11:17:39.358> Connection established.
d12 11:17:39.427> Tx failed, gasPrice = 19  :(
d12 11:17:39.503> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: n
once too low
```

After this the miner got stuck in a loop saying "nonce too low".  Restarting the miner fixed the loop, but gas bidding still failed and the solution was lost.

Bit of a shot in the dark, but is the miner multithreaded, and if so, can transactions be sent sequentially so as to avoid duplicate nonce/transactions?

**Comment by treevez** [comment]
As an update, I tried with NonceGeneration=Random and had the same problem, each time happening at 19 gas.  However, this time 19 gas succeeded, as shown by logs below:

```
d12 12:33:03.331> Trying again in 5 seconds ...
d12 12:33:08.399> Connecting to node at 127.0.0.1:8545 ...
d12 12:33:08.773> Connection established.
d12 12:33:08.846> Tx succeeded, gasPrice = 19  :)
d12 12:33:08.928> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: n
once too low
d12 12:33:09.004> Trying again in 5 seconds ...
d12 12:33:14.074> Connecting to node at 127.0.0.1:8545 ...
d12 12:33:14.434> Connection established.
d12 12:33:14.510> Tx succeeded, gasPrice = 19  :)
d12 12:33:14.569> An error occurred communicating with the node : Exception -32000 : Configuration file was not found: n
once too low
```

The miner was still put into an error loop, but the solution was successful.  I do not believe this had anything to do with NonceGeneration=Random, but that 19 gas just happened to be enough to win the bid.

**Comment by mining-visualizer** [comment]
Ha ha.  This could be interesting.  :)  I was wondering what was going to happen when everybody started bidding against each other using the same bidding algorithm.  When I was mining no one was actively bidding against me, so the most I ever saw was one bid-up event.  It is going to be difficult for me to reproduce that since I only have 2 gpus so I'm pretty much out of the game for solo mining, but I will study the code and try to come up with something.

The NonceGeneration has no effect ... it is a carry-over from the old ethminer.

Are you cutting some of the lines out of the log?  No problem if you are.  But normally it displays the bids coming in from the other miners as well.  I'll have to check my code on that.  If you want to keep that private thats fine.

**Comment by mining-visualizer** [comment]
Also, if you have multiple mining rigs,  make sure you set up a  ChallengeFolder to make sure you don't end up bidding against one of your other miners.

**Comment by mining-visualizer** [comment]
Also, just a comment about how the bidding works.  When the miner wants to bid up, it submits the same tx with the same nonce but a higher gas price.  According to my research, that is the way to go about it.  You could increase the nonce, but then you would have this trail of txs that would all fail and cost you unnecessary gas.

**Comment by mining-visualizer** [comment]
It just occurred to me, the log info you posted was screen log data.  Could you instead post log data from %appdata%\ethminer\log.txt.  That contains more detailed info.

**Comment by treevez** [comment]
Heres the log.txt from the first event.

```
d12 11:15:57.475> New challenge :
d12 11:17:16.011> Solution found; Submitting to node ...
d12 11:17:16.097> Tx hash :, gasPrice : 15
d12 11:17:16.797> Miner submitted tx. gasPrice=15, challenge=
d12 11:17:17.934> Adjusting gas price to 17, new tx hash=
d12 11:17:17.941> Miner submitted tx. gasPrice=17, challenge=
d12 11:17:19.347> Adjusting gas price to 19, new tx hash=
d12 11:17:27.111> New challenge :
d12 11:17:39.427> Tx failed, gasPrice = 19  :(
d12 11:17:45.086> Tx failed, gasPrice = 19  :(
d12 11:17:52.093> Tx failed, gasPrice = 19  :(
```

The miner gets stuck in the gasPrice = 19 loop right here until manually restarted.

Here's the log.txt from the succeeding event.

```
d12 12:29:01.396> New challenge :
d12 12:30:03.859> Solution found; Submitting to node ...
d12 12:30:03.952> Tx hash :, gasPrice : 15
d12 12:30:04.211> Miner submitted tx. gasPrice=15, challenge=
d12 12:30:05.348> Adjusting gas price to 17, new tx hash=
d12 12:30:05.355> Miner submitted tx. gasPrice=17, challenge=
d12 12:30:06.744> Adjusting gas price to 19, new tx hash=
d12 12:30:26.214> New challenge :
d12 12:30:26.406> Tx succeeded, gasPrice = 19  :)
d12 12:30:33.038> Tx succeeded, gasPrice = 19  :)
```

The miner successfully submits the solution, but gets stuck at gasPrice = 19

**Comment by mining-visualizer** [comment]
I think I found the problem ... the miner is bidding against itself!  I originally had my own mining addresses hard coded in the program to avoid bidding against myself, but when I released the program I deleted those but forgot to change that to the users address.  Just disable bidding for now, I'll have a release out in a few hours.

**Comment by treevez** [comment]
Ha! Happy it was a quick find.  I'll continue using infura for now and wait on that release, thanks for the ongoing effort.  I've removed identifying info from my logs and will close this once I get a chance to test the bidding fix.

**Comment by mining-visualizer** [comment]
Ok, v2,0.3 is available now on the Releases page.  Give it a try.

**Comment by treevez** [comment]
Testing now, will report back when I find a solution, shouldn't take more than a few hours (hopefully).

**Comment by treevez** [comment]
Miner no longer bids against itself.  If I notice any problems with bidding (all solutions have been accepted immediately since updating so far), I'll open a new issue.
